src/sdk/pynni/nni/compression/torch/utils/mask_conflict.py

```diff
@@ -221,73 +221,81 @@ def __init__(self, masks, model=None, dummy_input=None, traced=None):
+            # channel_masks is a list, each element is None or a vector, for example:
+            # [[0, 1, 1, 0, 0], [0, 0, 1, 1, 0], None], None means no channel
+            # is pruned.
+            channel_masks = []
             for name in dset:
-                if name not in self.masks:
-                    # this layer is not pruned
-                    all_pruned = False
-                    continue
-                w_mask = self.masks[name]['weight']
-                if out_channels is None:
-                    out_channels = w_mask.size(0)
-                shape = w_mask.size()
-                count = np.prod(shape[1:])
-                all_ones = (w_mask.flatten(1).sum(-1) == count).nonzero().squeeze(1).tolist()
-                all_zeros = (w_mask.flatten(1).sum(-1) == 0).nonzero().squeeze(1).tolist()
-                if len(all_ones) + len(all_zeros) < w_mask.size(0):
-                    # In fine-grained pruning, there is no need to check
-                    # the shape conflict
-                    _logger.info('Layers %s using fine-grained pruning', ','.join(dset))
-                    fine_grained = True
-                    break
-                channel_remain.update(all_ones)
-                _logger.debug('Layer: %s ', name)
-                _logger.debug('Original pruned filters: %s', str(all_zeros))
-            # Update the masks for the layers in the dependency set
-            if fine_grained or out_channels is None:
-                # if use the fine-grained pruner or all the layers in
-                # this dependency set are not pruned
+                if name in self.masks:
+                    if len(self.masks[name]['weight'].size()) == 4: # Conv
+                        channel_masks.append((self.masks[name]['weight'].abs().sum(sum_idx) != 0).int())
+                    elif len(self.masks[name]['weight'].size()) == 2: # Linear
+                        channel_masks.append((self.masks[name]['weight'].abs().sum(0) != 0).int())
+                    elif len(self.masks[name]['weight'].size()) == 1: # BatchNorm2d
+                        channel_masks.append((self.masks[name]['weight']).int())
+                    else:
+                        raise RuntimeError('unsupported mask shape')
+                else:
+                    # no mask means not pruned, equivlent to full masks
+                    channel_masks.append(None)
+
+            if all(x is None for x in channel_masks):
                 continue
-            if not all_pruned:
-                # if some layer are not pruned at all
-                # then all the layers in this dependency set
-                # cannot be pruned due to the shape dependency.
-                channel_remain.update(range(out_channels))
-            ori_channels = 0
+            num_channels_list = [len(x) for x in channel_masks if x is not None]
+            # number of channels in same set should be identical
+            assert len(set(num_channels_list)) == 1
```

Comment:
it is possible that the numbers of kept channels of multiple convs are different? because each layer's sparsity is specified by users

Reply:
This is total number of channels on specified dimension.  should be same.